Scripts/Deadmines: New script for Edwin VanCleef

Teppic1 opened the pull request:
**Changes proposed:**

Move VanCleef from SAI to C++ to correct various evade/boss/guard behaviour. Also fix issues with the existing scripting.

- Blackguards are now summoned properly as part of the encounter
- Blackguards and VanCleef will despawn/respawn on evade
- Fixed scripted actions/speech on health
- Removed Blackguards from world DB
- Corrected VanCleef's damage

**Target branch(es):** 3.3.5

- [x] 3.3.5

**Tests performed:** (Does it build, tested in-game, etc.)

Built and tested. 

**Known issues and TODO list:**

None as yet.

Teppic1 committed:
Scripts/Deadmines: New script for Edwin VanCleef

- Blackguards are now summoned properly as part of the encounter
- Blackguards will despawn on evade
- Fixed scripted actions/speech on health
- Removed Blackguards from world DB
- Corrected VanCleef's damage

jackpoz commented:
> Move VanCleef from SAI to C++

lol half of the PRs are about moving a script from a system to another xD

jackpoz commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+            _DespawnAtEvade();
+        }
+
+        void Reset() override
```
should this reset the _phase too ?

Teppic1 replied:
After evade he'll respawn and so it gets reset anyway.

Aokromes commented:
well, bosses are supposed to be c++ quest etc if possible SAI :P

jackpoz commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+            _phase = PHASE_NORMAL;
+        }
+
+        void Initialize()
```
as a standard we use Initialize() only to set the class fields to a default value, called in the ctor and in Reset(). Maybe this code fits better in Reset().

Keader replied:
yeap, it should be in reset

Teppic1 replied:
Ok, I put it there mainly because the compiler complained if I didn't :)

jackpoz commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+        {
+            for (auto BlackguardPosition : BlackguardPositions)
+                DoSummon(NPC_BLACKGUARD, BlackguardPosition, 60000,
+                        TEMPSUMMON_CORPSE_TIMED_DESPAWN);
```
please keep all DoSummon() parameters on 1 single line, it's a short line anyway

jackpoz commented:
The PR overall looks good, there are some minor things here and there. I dunno if we should handle the "from 100% hp to 20% hp" in a particular way.

Keader commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+
+            switch (_phase)
+            {
+            case PHASE_66_PCT:
```
identation

Keader commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+
+        void _SummonBlackguards()
+        {
+            for (auto BlackguardPosition : BlackguardPositions)
```
why auto?

Teppic1 replied:
The variable's only used for the loop, so why not? :)

Keader replied:
Naah, put position there
we only use auto in some iterators case

Keader commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+            });
+        }
+
+    protected:
```
private

Teppic1 replied:
Changed to all your suggestions, plus used the new model.

Keader commented:
What you say about use new script model?
Like in: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/boss_illidan.cpp#L1360-L1410

ccrs commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+                        TEMPSUMMON_CORPSE_TIMED_DESPAWN);
+        }
+
+        void _SwitchPhase(uint8 Phase)
```
"_" points that this is private, make it private then

ccrs commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+            Initialize();
+        }
+
+        void UpdateAI(uint32 diff) override
```
no need to override this, its the class default behaviour IIRC

offl commented:
Perhaps everything here can be done via SAI >_< Maybe Killyana can confirm that
I also can't undersand the manipulating with health, why he should say or cast only between 25-50%?

Aokromes commented:
^^ https://github.com/TrinityCore/TrinityCore/pull/20736#issuecomment-340165004

Teppic1 commented:
Also if it's left to SAI when he evades he will just run back to his spawnpoint, which isn't how bosses are supposed to behave. I originally tried with SAI using a spawn group for his initial guards but had multiple issues with them not despawning properly, him not aggroing them, etc.

Teppic1 committed:
Scripts/Deadmines: Remove Initialize() code and place into Reset()

Teppic1 committed:
Scripts/Deadmines: Remove unnecessary override

Teppic1 committed:
Scripts/Deadmines: formatting

ForesterDev commented:
I think, is better to use CreatureGroups for guards instead of hardcoding summon

Teppic1 committed:
Scripts/Deadmines: protected -> private

ccrs commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,170 @@
+            _DespawnAtEvade();
+        }
+
+        void UpdateAI(uint32 diff)
```
uh? just remove the whole thing

ForesterDev commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,170 @@
+            _DespawnAtEvade();
+        }
+
+        void UpdateAI(uint32 diff)
```
Why tou override this function? It's the same implementation as https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/ScriptedAI/ScriptedCreature.cpp#L542

Teppic1 replied:
I was following what I saw in existing scripts, but it's removed now.

ForesterDev commented:
Also, you should use new script model according https://github.com/TrinityCore/TrinityCore/commit/fb87ac8e8da3d215d4988b8c2fa25d7a98cf9ee0#diff-5d3914b5fea050ec9beba58cf348400d

Teppic1 committed:
Scripts/Deadmines: Properly remove unnecessary override

Teppic1 committed:
Scripts/Deadmines: change to new script model (hopefully)

Teppic1 commented:
Tested without the ``UpdateAI()`` override and he doesn't do any events any more. With it back in it works again.

Teppic1 committed:
Scripts/Deadmines: previous commit stopped events working

Teppic1 commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,176 @@
+        {
+            scheduler.Schedule(Seconds(1), [this](TaskContext task)
+            {
+                if (_phase == PHASE_NORMAL && HealthBelowPct(66))
```
At the moment everything "skipped" will be queued up and executed, I'm not sure if that's desirable or not.

Keader commented:
> Tested without the UpdateAI() override and he doesn't do any events any more. With it back in it works again.

It's because you using TaskScheduler to handle with events.
You have 2 choices here: Or keep your UpdateAI() or make a new PR adding scheduler.Update(diff) here: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/ScriptedAI/ScriptedCreature.cpp#L547

Teppic1 commented:
@Keader It's possible just to add the line ``scheduler.Update(diff);`` to ``BossAI::UpdateAI()`` and it not break tons of stuff? I tried it and removed the ``UpdateAI()`` from my code and it was fine as far as that was concerned.

Keader commented:
will not, you just making generic UpdateAI update all events based in TaskScheduler
It should be default behavior, idk is not present in current revs...
Is better add this line in a new PR

Keader commented:
Hmm
Thinking better, just restaure your UpdateAI.
We always use EventMap in events, update TaskScheduler too is unecessary.

nawuko commented:
I would rather not recommend to use licensed files (only remove the header). Not only can this be troublesome from the legal site for you, it also is a huge problem for OSS repositories like TrinityCore to get a DMCA takedown.

Keader commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,160 @@
+struct boss_vancleef : public BossAI
+{
+    public:
+        boss_vancleef(Creature* creature) : BossAI(creature, 0) { }
```
magic number
should have a DATA_BOSS_VANCLEEF or something like that in deadmines.h
Btw deadmines.h should be called Deadmines.h

Teppic1 replied:
Yeah I wondered about that but since there's only one boss with BossAI I didn't know if it would be necessary.

Keader replied:
yes, is

Teppic1 replied:
That's fixed now. I added Mr Smite since he's currently the only other boss scripted, but doesn't use BossAI (that probably needs to be rewritten too)

Teppic1 commented:
@nawuko I spent hours looking through multiple files (including one you wrote) to learn TC's system, it's not simply your file with a header change.

Commit:
Scripts/Deadmines: remove magic number, rename deadmines.h

Commit:
Scripts/Deadmines: update #include directives

Teppic1 commented:
@tkrokli  - this was already done, the change below was local only

> Hmm
> Thinking better, just restaure your UpdateAI.
> We always use EventMap in events, update TaskScheduler too is unecessary.

Keader commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,160 @@
+            });
+        }
+
+        void ScheduleTasks() override
```
What Nawuko suggested is true.
This phase events should be handled in DamageTaken hook instead a 1 sec event

Teppic1 replied:
I can rewrite that, but it is wasting my time if it'll never get accepted due to false accusations.

tkrokli commented:
>  @tkrokli - this was already done, the change below was local only

Well, my reason for changing the label was more based on the fact that it looked like there are still some details to sort out. But if it will make you feel better, I can change the label back while still leaving the decision on when to merge up to the main developer members.

Teppic1 commented:
I've addressed all the comments, except the last one which I'm waiting for a response on.

Keader commented:
@nawuko has some prof of "stolen work" ?

tkrokli commented:
Would be a good idea to be very clear about the details for such accusations.
What has been said so far seems cryptic and/or obscure to me.

Teppic1 commented:
I can provide a copy of the file he is claiming is just a whitespace change (it's licensed under the GPL2 after all). There are many similarities since this is a pretty simple script, not a raid boss, and so it's naturally overriding the same functions etc.

The situation is that we were working together on a TC derived emulator, and so he distributed his code to me to run on the server I was paying for. If I wanted to upload something of his I'd not choose some simple Deadmines boss that I'm trying to learn TC scripting by doing myself.

nawuko commented:
> I can provide a copy of the file he is claiming is just a whitespace change (it's licensed under the GPL2 after all).

Troublesome from a legal standpoint, and also wrong.

> There are many similarities since this is a pretty simple script

And so you end up with same function names, same timing, same percentages (which was made up by me and is not 100% true from retail). Even the same file structure (where function are withing in class). Even small errors are from me are included.

> If I wanted to upload something of his I'd not choose some simple Deadmines boss that I'm trying to learn TC scripting by doing myself.

So lets upload DBC files directly from the WoW Client, we are all here to learn something right?

 **No** - this is not how this works, you have to respect copyright even if its "for learning". The reason why you do it dosen't justify the action.

Also, like i said i have no problem to get inspired, but i have a problem with using my copyrighted work (and offering to distribute it). Also you just **admitted** in a Private Reddit message that you **took parts from my work**. 

So to close this, since the script has come a long way and is now far from the original, you can continue your work. But i highly discourage you from taking parts from my work again **without permission.** 

Also to add to the conversation, i would remove the ::EnterEvadeMode code section (the base class handles the despawn of summons.). Since the behavior from _DespawnAtEvade can A) cause some respawn issues atm. B) Is not blizzlike at all for classic/tbc dungeon bosses. If you don't use _DespawnAtEvade their is no reason to summon the first 2 guards (in the ::Reset overwrite) via script. Since this part is also not blizzlike, you where able to kill the guards during an encounter -> wipe and not fight them again. With this implementation you have to fight them every time the boss resets.

Teppic1 commented:
Here is a copy of the file this ridiculous drama is about:

https://pastebin.com/mMs91NE8

And here is the file in this PR:

https://github.com/Teppic1/TrinityCore/blob/815aa0e4610646270291c5ad0bc7d12f3d82c3b8/src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp

Some things are identical, since they are identical in 1000 other TC scripts too. Some things are not: the guard despawn in this file doesn't work, it does in mine; I use the Blizzard spell 5200 to summon guards in combat, this file uses DoSummon(). There are lots of other changes as can be clearly seen.

The part I used from the above file was the phase section, which seemed perfectly suitable and mostly trivial, since it's just if/else. It's natural to see small sections of code in any file you view and reuse them, which is part of the reason the GPL exists.

I would appreciate any further drama be directed to me instead of made public in this extremely immature and offensive way.

Wyreth commented:
All this talk about copyright is quite discouraging, since we're all mostly here to learn and help improve on something we all love (I guess?).

I'm not sure if a bunch of variables and API calls (which is what TC scripting is, after all) fall into the copyright laws. But that's beyond the point.

nawuko commented:
> Some things are identical, since they are identical in 1000 other TC scripts too

First of all, you are admitting that you took parts. Second of all even the variable names are the same for the function _SummonGuards (which has also the same name, and is on the same place in the class)

> The part I used from the above file was the phase section

Clearly not the case, explained above and in the above posting.

> I would appreciate any further drama be directed to me instead of made public in this extremely immature and offensive way.

Unable to read? Also you are the guy that started with insults in the private reddit message, not me.

@Wyreth Yes and no, GPL is really the wrong license for TC (and they know it, but its impossible to change that since you would need an agreement from everyone that has contributed [if the code from them is still in there]). Since a server does not distribute any form of binary many requirements from GPL are not valid.

Teppic1 commented:
@Wyreth In a script this simple it's little more than a set of overrides that are used identically in every other current TC script that needs to perform the same actions. It made sense to choose a boss like this to work on rather than something that does actually involve non-trivial logic.

Kittnz commented:
Will be reopened tomorrow.

Teppic1 committed:
Scripts/Deadmines: Various improvements:
 - Move NPC_BLACKGUARD enum to Deadmines.h since Mr Smite uses them
 - Rename PHASE enum because the word normal is apparently copyrighted
 - Remove ScheduleTasks() and UpdateAI() in favour of simpler DamageTaken()

Teppic1 committed:
Scripts/Deadmines: make Travis happy

ccrs commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,142 @@
+
+        void DamageTaken(Unit* /*attacker*/, uint32& /*damage*/) override
+        {
+            if (_phase == PHASE_ONE && HealthBelowPct(66))
```
HealthBelowPctDamaged is better
Also I'd change the order of the checks, example scenario: a level 80 uses a spell that brings the boss to 5% health, which phase should be triggered? One or Last?

ccrs replied:
And also, DamageTaken is NOT a standard in TC for phase switching
I personally prefer to check it on update since DamageTaken can be called multiple times on the same update tick, from different damage incomes, giving the chance to multiple same triggered actions

jackpoz replied:
@ccrs I wondered the same about hp % at https://github.com/TrinityCore/TrinityCore/pull/20736#issuecomment-340165344

Keader replied:
you can control multiple calls using a variable, i dont see any problem about it.
But i agree with you, is better change the order.

ariel- replied:
A little late but IIRC blizz uses update to check hp

Teppic1 committed:
Replace TDB guard positions with sniffed data

Teppic1 commented:
I thought it might be better just to get rid of the phase stuff, do a flag check every time for the guard summons (since this is max priority) and then fall through to the speech checks, I guess in reverse order from what it currently is.

Teppic1 committed:
Scripts/Deadmines: Remove phase code and replace with flags, reverse checks

Aokromes committed:
Rename Deadmines.h to deadmines.h

Aokromes committed:
Update instance_deadmines.cpp

Aokromes committed:
Update boss_mr_smite.cpp

Aokromes committed:
Update deadmines.cpp

Aokromes committed:
Update boss_vancleef.cpp

Teppic1 committed:
Use TDB coords for guards, values taken from sniff are incorrect

jackpoz commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,128 @@
+        {
+            BossAI::Reset();
+
+            _guardsCalled = false;
```
please initialize all class field members in the constructor too to avoid static analysis issues

tkrokli replied:
Looks like this was done in the following commit (https://github.com/TrinityCore/TrinityCore/pull/20736/commits/7664213442fd7e97259d5c23c78cae38b1065c1d).
Does the PR look good enough for your approval now? :-)

Teppic1 replied:
Yeah I forgot to comment here that I'd addressed that (all the other comments made have been addressed too).

jackpoz reviewed (changes requested):
The script looks all good except a couple of class fields that need to be initialized in the constructor.
Other than that, is the PR ready to be merged ?

Teppic1 committed:
Initialise private members

Teppic1 committed:
Fix warning

Aokromes commented on src/server/scripts/EasternKingdoms/Deadmines/boss_vancleef.cpp:
```diff
@@ -0,0 +1,128 @@
+ */
+
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
```
move the last include between the 2 script* like on https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/IcecrownCitadel/boss_lord_marrowgar.cpp

Teppic1 committed:
Move include as requested

jackpoz committed:
Merge branch '3.3.5' into vancleef

jackpoz committed:
Rename sql

jackpoz commented:
Thanks for the PR
